README.md
# BBC Games - Genie Starter Pack

* [Genie - Getting Started](./docs/getting-started.md)
* [Genie Component Delivery Definitions](./docs/genie-component-delivery-definitions.md)
* [Developing a Gameplay Component](./docs/development/gameplay-component.md)
* [Creating a Theme](./docs/theming.md)
* [Tech Review Tool](./docs/tech-review-tool.md)
* [Testing](./docs/testing.md)

## Gameplay Development Specifics

* [Code Restrictions](./docs/development/code-restrictions.md)
* [Games Messaging Interface (GMI)](./docs/development/gmi.md)
* [Using Local Storage](./docs/development/data-storage.md)
* [Game Settings](./docs/development/settings.md)
* [The Signal Bus](./docs/development/signal-bus.md)
* [Settings](./docs/development/settings.md)
* [Accessibilify](./docs/development/accessibilify.md)
* [Brim](./docs/development/brim.md)
* [Phaser Terminology](./docs/development/phaser-terminology.md)
* [Stats](docs/development/stats.md)
* [States](docs/development/stats.md)

## Genie Screens
* [Results](./docs/development/results-screen/results-start.md)
* [Narrative](./docs/development/narrative-screen.md)
* [Select](./docs/development/select-screen.md)


CHANGELOG.md

| Version |                                          Change                                           |
|---------|-------------------------------------------------------------------------------------------|
| 0.0.21  | Bump Genie Core to 3.9.0 |
| 0.0.20  | Bump Genie Core to 2.0.8 |
| 0.0.19  | Bump Genie Core to 2.0.7 |
| 0.0.18  | Bump Genie Core. |
| 0.0.17  | Send level ID in level~complete stat.|
| 0.0.16  | Remove achievement audio.|
| 0.0.15  | Add level id stats docs. Add validate docs and script.|
| 0.0.14  | Bump Genie Core. Update Docs. |
| 0.0.13  | Add achievements documentation. |
| 0.0.12  | Add achievements. |
| 0.0.11  | Add achievement files. |
| 0.0.10  | Include a numeral in the result score so score is reported in stats. |
| 0.0.9   | Prune scripts, add sourcemaps support to start:pack. |
| 0.0.8   | Bump Genie core and include example stat calls. |
| 0.0.7   | Bump Genie core. |
| 0.0.6   | Bump Genie core. Updates for Babel 7 |
| 0.0.5   | Bump version of Genie core. |
| 0.0.4   | npm audit fix. Bump version of Genie core. change build-watch script to build:watch for consistency. |
| 0.0.3   | Correct webpack dev server calls to create an output folder |


src/main.js
import { Home } from "/node_modules/genie/src/components/home.js";
import { Select } from "/node_modules/genie/src/components/select/select-screen.js";
import { Shop } from "/node_modules/genie/src/components/shop/shop.js";
import { HowToPlay } from "/node_modules/genie/src/components/how-to-play.js";
import { Pause } from "/node_modules/genie/src/components/overlays/pause.js";
import { Game } from "./components/game.js";
import { Results } from "/node_modules/genie/src/components/results/results-screen.js";
import { Narrative } from "/node_modules/genie/src/components/narrative.js";
import { startup } from "/node_modules/genie/src/core/startup.js";
import { settingsChannel } from "/node_modules/genie/src/core/settings.js";
import { eventBus } from "/node_modules/genie/src/core/event-bus.js";
/*
    IMPORTANT NOTE FOR AGENCIES

    Do not arbitrarily change the above built in Genie screens for copies with your own modifications
    without prior consent from a BBC project manager
*/

// Setup for BBC settings control
const settings = {
    pages: [
        {
            title: "Custom Settings",
            settings: [
                {
                    key: "custom1",
                    type: "toggle",
                    title: "Custom setting",
                    description: "Description of custom setting",
                },
            ],
        },
    ],
};

// Example of responding to custom game settings
eventBus.subscribe({
    channel: settingsChannel,
    name: "custom1",
    callback: value => {
        console.log("Custom 1 setting changed to " + value);
    },
});

// Additional game options can be passed to phaser here
const gameOptions = {
    //pixelArt: true,
};

const screens = {
    home: {
        scene: Home,
        routes: {
            debug: "debug",
            //Example of custom routing function
            next: scene => {
                scene.navigate("narrative");
            },
        },
    },
    narrative: {
        scene: Narrative,
        routes: {
            next: "character-select",
        },
    },
    "character-select": {
        scene: Select,
        routes: {
            next: "level-select",
            home: "home",
        },
    },
    "level-select": {
        scene: Select,
        routes: {
            next: "game",
            home: "home",
        },
    },
    game: {
        scene: Game,
        settings: {
            physics: {
                default: "arcade",
                arcade: {},
            },
        },
        routes: {
            next: "results",
            home: "home",
            restart: "game",
        },
    },
    results: {
        scene: Results,
        routes: {
            continue: "level-select",
            restart: "game",
            home: "home",
        },
    },
    ...Shop({ key: "shop", routes: {} }),
    // Overlays
    "how-to-play": {
        scene: HowToPlay,
        routes: {
            home: "home",
        },
    },
    pause: {
        scene: Pause,
        routes: {
            home: "home",
            select: "character-select",
        },
    },
};

startup({ screens, settings, gameOptions });


build-scripts/webpack.config.js
var path = require("path");

// Phaser webpack config
var phaserModule = path.join(__dirname, "../node_modules/phaser-ce/");
var phaser = path.join(phaserModule, "build/custom/phaser-split.js");
var pixi = path.join(phaserModule, "build/custom/pixi.js");
var p2 = path.join(phaserModule, "build/custom/p2.js");

module.exports = {
    mode: "production",
    performance: { hints: false },
    entry: ["babel-polyfill", "pixi", "p2", "phaser", "webfontloader", path.resolve("src/main.js")],
    output: {
        path: path.resolve("output"),
        publicPath: "output",
        filename: "main.js",
    },
    module: {
        rules: [
            { test: /\.js$/, use: ["babel-loader"], include: path.resolve("src") },
            { test: /\.js$/, use: ["babel-loader"], include: path.resolve("node_modules/genie/src") },
            { test: /pixi\.js/, use: ["expose-loader?PIXI"] },
            { test: /phaser-split\.js$/, use: ["expose-loader?Phaser"] },
            { test: /p2\.js/, use: ["expose-loader?p2"] },
            { test: /webfontloader\.js/, use: ["expose-loader?WebFont"] },
        ],
    },
    resolve: {
        alias: {
            phaser: phaser,
            pixi: pixi,
            p2: p2,
        },
    },
};


src/components/balloon-pump-game.js
import { Screen } from "../../node_modules/genie/src/core/screen.js";

export class BalloonPumpGame extends Screen {
    constructor() {
        super();
    }

    preload() {
        this.keyLookup = this.layoutFactory.keyLookups[this.game.state.current];
    }

    create() {
        const centerX = this.game.world.centerX;
        const centerY = this.game.world.centerY;
        this.theme = this.context.config.theme[this.game.state.current];

        this.addBackground();
        this.createTimer(centerX);
        this.createGameFurniture(centerX, centerY);
        this.createBalloon(centerX, centerY);

        this.layoutFactory.addLayout(["home", "pause", "audioOff", "settings"]);
    }

    render() {
        if (this.timer.running) {
            this.timeLeftText.setText(this.getTimeLeftString());
        }
    }

    update() {
        if (this.balloonButton.width >= 240) {
            this.next({ transient: { resultsData: "Finished with " + this.getTimeLeft() + " seconds left!" } });
        } else if (this.balloonButton.width > 74 && this.balloonButton.height > 122) {
            this.balloonButton.width -= 1;
            this.balloonButton.height -= 1;
        }
    }

    gameLost() {
        this.next({ transient: { resultsData: "Game over - You lost!" } });
    }

    addBackground() {
        const backgroundImage = this.game.add.image(0, 0, this.keyLookup.background);
        return this.layoutFactory.addToBackground(backgroundImage);
    }

    createTimer() {
        this.timer = this.game.time.create();
        this.timerEvent = this.timer.add(Phaser.Timer.SECOND * 15, this.gameLost, this);
        this.timeLeftText = this.game.add.text(-270, 230, this.getTimeLeftString(), this.theme.timer.style);
        this.layoutFactory.addToBackground(this.timeLeftText);
        this.timer.start();
    }

    createGameFurniture() {
        const pumpSprite = this.game.add.sprite(-70, 190, this.keyLookup.pump);
        this.layoutFactory.addToBackground(pumpSprite);

        const text = this.game.add.text(0, -170, this.theme.text.content, this.theme.text.style);
        this.layoutFactory.addToBackground(text);
    }

    createBalloon() {
        this.balloonButton = this.game.add.button(-10, 110, this.keyLookup.balloon, this.balloonTapped, this, 2, 1, 0);
        this.balloonButton.width = 74;
        this.balloonButton.height = 122;
        this.layoutFactory.addToBackground(this.balloonButton);
        this.balloonButton.anchor.setTo(0.5, 1);
    }

    balloonTapped() {
        this.balloonButton.width += 20;
        this.balloonButton.height += 20;
    }

    getTimeLeft() {
        return Math.round((this.timerEvent.delay - this.timer.ms) / 1000);
    }

    getTimeLeftString() {
        return this.formatTime(this.getTimeLeft());
    }

    formatTime(secs) {
        const minutes = 0 + Math.floor(secs / 60);
        const seconds = 0 + (secs - minutes * 60);
        return "Time left " + this.zeroPad(minutes) + ":" + this.zeroPad(seconds);
    }

    zeroPad(numberToPad) {
        const zero = 2 - numberToPad.toString().length + 1;
        return Array(+(zero > 0 && zero)).join("0") + numberToPad;
    }
}


src/components/click-progression-game.js
import { Screen } from "../../node_modules/genie/src/core/screen.js";
import { gmi } from "../../node_modules/genie/src/core/gmi/gmi.js";

const calculateAchievements = remaining => {
    gmi.achievements.set({key: "first_win"});
    if (remaining >= 12) {
        gmi.achievements.set({key: "fast_12"});
    }
    if (remaining >= 10) {
        gmi.achievements.set({key: "fast_10"});
    }
    if (remaining >= 8) {
        gmi.achievements.set({key: "fast_8"});
    }
    if (remaining >= 6) {
        gmi.achievements.set({key: "fast_6"});
    }
    if (remaining >= 4) {
        gmi.achievements.set({key: "fast_4"});
    }
    const gameData = gmi.getAllSettings().gameData;
    if (!gameData.timesCracked) {
        gmi.setGameData("timesCracked", 0);
    }
    let timesCracked = gameData.timesCracked;
    timesCracked = timesCracked + 1;
    gmi.setGameData("timesCracked", timesCracked);
    gmi.achievements.set({key: "many_wins", progress: timesCracked});
    gmi.achievements.set({key: "many_wins_2", progress: timesCracked});
    gmi.achievements.set({key: "many_wins_3", progress: timesCracked});
}

const zeroPad = numberToPad => {
    const zero = 2 - numberToPad.toString().length + 1;
    return Array(+(zero > 0 && zero)).join("0") + numberToPad;
}

const formatTime = secs => {
    const minutes = Math.floor(secs / 60);
    const seconds = secs - minutes * 60;
    return "Time left " + zeroPad(minutes) + ":" + zeroPad(seconds);
}

export class ClickProgressionGame extends Screen {

    create() {
        this.theme = this.context.config.theme[this.scene.key];

        this.addBackground();
        this.createTimer();
        const gameText = this.add.text(this.theme.text.position.x, this.theme.text.position.y, this.theme.text.content, this.theme.text.style);
        gameText.setOrigin(0.5);

        const character = this.transientData["character-select"];
        const level = this.transientData["level-select"];
        this.selectedGameButton = level.choice.title.match(/(\d+)/)[0];
        this.selectedGameCharacter = character.choice.title.match(/(\d+)/)[0];
        
        const buttonPos = this.theme.gameButton.position;
        const positions = [
            { x: buttonPos.x, y: buttonPos.y - 280 },
            { x: buttonPos.x + 210, y: buttonPos.y - 280 },
            { x: buttonPos.x - 210, y: buttonPos.y - 280 },
            { x: buttonPos.x, y: buttonPos.y - 140 },
            { x: buttonPos.x + 210, y: buttonPos.y - 140 },
            { x: buttonPos.x - 210, y: buttonPos.y - 140 },
        ]

        let buttonPosition = 0;
        buttonPosition = level.choice.title.match(/(\d+)/)[0];
        const buttonPositions = positions.slice(0, buttonPosition);

        this.gameButtons = buttonPositions.map(pos => this.createGameButton(pos.x, pos.y))

        this.setLayout(["pause"]);

        gmi.setGameData("characterSelected", this.transientData["character-select"]);
        console.log("Data saved to GMI:", gmi.getAllSettings().gameData); // eslint-disable-line no-console
        gmi.sendStatsEvent("level", "start", { source: level.choice.title });
    }

    update() {
        this.timeLeftText.setText(this.getTimeLeftString());
    }

    gameButtonClicked(button) {
        if ( button.getData("timesClicked") < 9 ) {
            button.data.values.timesClicked += 1;
            button.setTexture("game." + "game_button_" + this.selectedGameCharacter + "_" + button.getData("timesClicked"));
            this.input.setHitArea(button);
        }

        const complete = this.gameButtons.every(button => button.getData("timesClicked") >= 9 );

        if (complete) {
            const level = this.transientData["level-select"];
            const remaining = this.getTimeLeft();
            const numberOfEggs = level.choice.title.match(/(\d+)/)[0];
            gmi.sendStatsEvent("level", "complete", {metadata:`SCO=[${remaining}]~LVR=[WIN]~SRC=[0]`, source: level.choice.title});
            calculateAchievements(remaining);
            this.transientData.results = { remaining, numberOfEggs };
            this.navigation.next();
        }
    }

    gameLost() {
        const level = this.transientData["level-select"];
        const remaining = 0;
        gmi.sendStatsEvent("level", "complete", {metadata:`SCO=[0]~LVR=[LOSE]~SRC=[0]`, source: level.choice.title});
        this.transientData.results = { remaining, numberOfEggs };
        this.navigation.next();
    }

    addBackground() {
        return this.add.image(0, 0, "game.background");
    }

    createTimer() {
        this.timerEvent = this.scene.scene.time.addEvent({delay: 15000, callback: this.gameLost, callbackScope: this });
        const timerText = this.timeLeftText = this.add.text(this.theme.timer.position.x, this.theme.timer.position.y, this.getTimeLeftString(), this.theme.timer.style);
        timerText.setOrigin(0.5);   
    }


    createGameButton(x, y) {
        const buttonImage = "game." + "game_button_" + this.selectedGameCharacter + "_0";
        console.log("buttonImage", buttonImage);
        const gameButton = this.add.image(x, y, buttonImage);
        gameButton.setInteractive().on("pointerdown", () => this.gameButtonClicked(gameButton));
        gameButton.setData("timesClicked", 0);
        gameButton.setOrigin(0.5);

        return gameButton
    }



    getTimeLeft() {
        return Math.round((this.timerEvent.delay - this.timerEvent.elapsed) / 1000);
    }

    getTimeLeftString() {
        return formatTime(this.getTimeLeft());
    }
}
